src/config.rs

```diff
@@ -230,6 +237,10 @@ impl Config {
             Network::Regtest => config.daemon_dir.push("regtest"),
         }
 
+        let blocks_dir = config.blocks_dir.unwrap_or(
+            default_blocks_dir(&config.daemon_dir)
```

Comment:
`unwrap_or_else` would avoid needless allocation in case `blocks_dir` is specified.

Reply:
Hmm yep but it takes a closure and would require to move config and borrow it (as we derive the default blocks dir from the daemon dir).

Reply:
Borrowing inside closure should be enough. Am I missing something?

Reply:
Hmm I'm still learning Rust so I was interested in trying this out. I couldn't come up with something smarter than another allocation to avoid `move`ing `config`:
```patch
diff --git a/src/config.rs b/src/config.rs
index 1bc4dda..a9975e9 100644
--- a/src/config.rs
+++ b/src/config.rs
@@ -235,8 +235,9 @@ impl Config {
             Network::Regtest => config.daemon_dir.push("regtest"),
         }
 
-        let blocks_dir = config.blocks_dir.unwrap_or(
-            default_blocks_dir(&config.daemon_dir)
+        let daemon_dir = config.daemon_dir.clone();
+        let blocks_dir = config.blocks_dir.unwrap_or_else(||
+            default_blocks_dir(&daemon_dir)
         );
 
         let cookie_getter =
```

Reply:
I don't see why `let blocks_dir = config.blocks_dir.unwrap_or(|| default_blocks_dir(&config.daemon_dir));` would be problematic, maybe I'll take a look at it some day.

Reply:
Because you try to borrow a moved value.

Reply:
Oh, I see now. It can be solved using additional variable, there's an RFC about making this case work out of the box. I think I will be able to make a PR soon.

Reply:
Hah, wanted to look at it now, but @romanz managed to do it before me (feel free to laugh imagining my puzzled look :D). If you're still interested in learning, you can take a look at his change: https://github.com/romanz/electrs/commit/521914a853088d5bf4276276df2648d644539f8f#diff-6d8a5e2104954a026de2bfe2b936aea4R239

Reply:
Thanks! :)
That's exactly what i came up with above: but in both cases there is a needless an allocation even when `blocks_dir` is specified (the original comment of this thread).

Reply:
Hmm? What needless allocation? He solved the allocation by changing `ok_or` to `ok_or_else`.

Reply:
Oh I read too fast: actually i was wrong he did `let daemon_dir = &config.daemon_dir;` while i came up with `config.daemon_dir.clone();` !